components/places/src/bookmark_sync/tests/payload_evolution.rs

```diff
@@ -0,0 +1,358 @@
+                vec![("device_type", "phone"), ("color", "blue")],
+            ),
+            remote_separator(guid('d'), 2, vec![("device_type", "desktop")]),
+            // We never change Query and Livemarks records, so we can't test roundtripping changes
+            // to those
+        ],
+        local_updates: vec![
+            LocalUpdate::BookmarkUpdate(title_change(guid('a'), "Doggies")),
+            LocalUpdate::BookmarkUpdate(title_change(guid('b'), "Kitties")),
+            LocalUpdate::BookmarkUpdate(title_change(guid('c'), "Cute Animals")),
+            LocalUpdate::BookmarkUpdate(pos_change(guid('d'), 3)),
+        ],
+        incoming_remote_records: vec![],
+        outgoing_unknown_fields: vec![
+            (guid('a'), vec![("device_type", "phone")]),
+            (guid('b'), vec![]),
+            (guid('c'), vec![("device_type", "phone"), ("color", "blue")]),
+            (guid('d'), vec![("device_type", "desktop")]),
+        ],
+    }
+    .run()
+}
+
+#[test]
+fn test_new_unknown_fields() {
+    // If we have new incoming remote records with new unknown fields, those should override the
+    // ones from the mirror table
+    RoundtripTest {
+        initial_remote_records: vec![
+            remote_bookmark(guid('a'), "Dogs", vec![("device_type", "phone")]),
+            remote_bookmark(guid('b'), "Cats", vec![("device_type", "desktop")]),
+        ],
+        local_updates: vec![
+            LocalUpdate::BookmarkUpdate(title_change(guid('a'), "Doggies")),
+            LocalUpdate::BookmarkUpdate(title_change(guid('b'), "Kitties")),
+        ],
+        incoming_remote_records: vec![remote_bookmark(
+            guid('a'),
+            "Dogs",
+            vec![("device_type", "mini-phone")],
+        )],
+        outgoing_unknown_fields: vec![
+            (guid('a'), vec![("device_type", "mini-phone")]),
+            (guid('b'), vec![("device_type", "desktop")]),
+        ],
+    }
+    .run()
+}
+
+#[test]
+fn test_reupload_after_reset() {
+    // After a reset, we re-upload all local records.  Those records should still have the unkown
+    // fields
+    RoundtripTest {
+        initial_remote_records: vec![
+            remote_bookmark(guid('a'), "Dogs", vec![("device_type", "phone")]),
+            remote_folder(
+                guid('b'),
+                "Animals",
+                vec!["A", "B"],
+                vec![("device_type", "phone"), ("color", "blue")],
+            ),
+            remote_separator(guid('c'), 2, vec![("device_type", "desktop")]),
+            // We re-upload query and livemark records after a reset, so test them here
+            // ... except for some reason apply_incoming() flags these as deleted and wants to sync
+            // outgoing tombstones for them
+            // remote_query(guid('d'), "dog search", "place:folder=123&excludeItems=1", vec![("device_type", "desktop")]),
+            // remote_livemark(guid('e'), vec![("device_type", "desktop")]),
```

Comment:
I really wanted to test this case, but I couldn't figure out this issue.  I'm hoping @mhammond knows what's going on here.

Reply:
You actually uncovered an intentional behavior! Livemarks were [removed in Firefox Desktop 64](https://support.mozilla.org/en-US/kb/feed-reader-replacements-firefox), and none of our other platforms ever supported them, so Sync deletes them.

There's a longer saga behind livemarks if you're curious! 😄 Forever ago, Desktop's Places stored them as folders with special annotations—the feed and site URLs—and real bookmark children for each feed item. Sync had a special record type for them, too, but otherwise treated them as regular folders. That scheme caused write and sync churn, though, so livemarks became empty folders with annotations, and the items were populated dynamically in the UI.

Some of our power users loved livemarks, but they never became a super popular feature, and couldn't justify their maintenance burden—so the Desktop team decided to remove them in 64. That release had a migration step that exported livemarks to OPML, then deleted them from Places.

We didn't want Sync to resurrect them—or turn them into folders—so we also added logic in [the merger](https://github.com/mozilla/dogear/pull/37) and [Desktop's mirror](https://bugzilla.mozilla.org/show_bug.cgi?id=1546035) to delete them.

TL;DR: Rust Places inherited the logic to delete livemarks from Desktop. Even though Fenix and iOS never supported livemarks, they'll still upload tombstones if they ever see one.

Since they're not a thing anymore, I think it's also totally OK not to support unknown fields for them—it doesn't hurt to, but it is extra code that none of our platforms should ever invoke.

I hope that helps!

Reply:
That's definitely helpful, thanks for the context here.

Do you know if queries are in the same category or are they still supported?